test/test_cpp_extensions_aot.py

```diff
@@ -133,6 +147,100 @@ def test_cuda_dlink_libs(self):
         test = cuda_dlink.add(a, b)
         self.assertEqual(test, ref)
 
+
+class TestPybindTypeCasters(common.TestCase):
+    """Pybind tests for ahead-of-time cpp extensions
+
+    These tests verify the types returned from cpp code using custom type
+    casters. By exercising pybind, we also verify that the type casters work
+    properly.
+
+    For each type caster in `torch/csrc/utils/pybind.h` we create a pybind
+    function that takes no arguments and returns the type_caster type. The
+    second argument to `PYBIND11_TYPE_CASTER` should be the type we expect to
+    receive in python, in these tests we verify this at run-time.
+    """
+    @staticmethod
+    def expected_return_type(func):
+        """
+        Our Pybind functions have a signature of the form `() -> return_type`.
+        """
+        # Imports needed for the `eval` below.
+        from typing import List, Tuple
+
+        return eval(re.search("-> (.*)\n", func.__doc__).group(1))
```

Comment:
Couldn't think of a reliable way to do this without `eval`. The poorly documented `pydoc.locate` works for regular classes, but not generics from `typing`.